### PR TITLE
[receiver/gitproviderreceiver] add Pull Request metrics

### PR DESCRIPTION
**Pull request by 4lch4**
**Description:**
This is the 3rd PR for the `gitproviderreceiver` that adds functionality for the following Pull Request metrics:

- `git.repository.pull_request.open.count`
- `git.repository.pull_request.open.time`
- `git.repository.pull_request.merged.count`
- `git.repository.pull_request.merged.time`
- `git.repository.pull_request.approved.time`

**Link to tracking Issue:** #22028 

**Testing:**
Basic tests included for new functionality.

**Documentation:**
Documentation has been updated to include information regarding the new metrics.

### CODE REVIEW

**Comment by 4lch4** [comment]
I resolved the merge conflicts so it should be ready to review :)

File: receiver/gitproviderreceiver/documentation.md
Comment (adrielp) [inline receiver/gitproviderreceiver/documentation.md:39]:
```suggestion
The amount of time it took for a pull request to be approved.
```

File: receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go
Comment (adrielp) [inline receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:210]: I'm thinking that it'd be better to convert to hours at the visualization/query level and the time/age value emitted from the receiver should be in seconds. The semantic conventions seem to leave it open in the case of time for metrics. 

> time - an instrument that measures passage of time should be called entity.time. For example, system.cpu.time with attribute state = idle | user | system | .... time measurements are not necessarily wall time and can be less than or greater than the real wall time between measurements.
> time instruments are a special case of usage metrics, where the limit can usually be calculated as the sum of time over all attribute values. utilization for time instruments can be derived automatically using metric event timestamps. For example, system.cpu.utilization is defined as the difference in system.cpu.time measurements divided by the elapsed time and number of CPUs.

This is primarily focused at time in the context of system metrics. I think just recording the real time in s/ms would be fine. Wouldn't mind knowing your thoughts on this @atoulme / @astencel-sumo

File: receiver/gitproviderreceiver/metadata.yaml
Comment (adrielp) [inline receiver/gitproviderreceiver/metadata.yaml:28]:
```suggestion
    description: The name of the Git repository
```

File: receiver/gitproviderreceiver/metadata.yaml
Comment (adrielp) [inline receiver/gitproviderreceiver/metadata.yaml:31]:
```suggestion
    description: The name of a Git branch
```

File: receiver/gitproviderreceiver/metadata.yaml
Comment (adrielp) [inline receiver/gitproviderreceiver/metadata.yaml:77]: When I read this, I think: "why would we want to know the age of a pull request that has been merged?"

File: receiver/gitproviderreceiver/metadata.yaml
Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml:28]: I believe it's just a matter of whether or not the owner of the repo is included. For example, in the case of the current repo we're in, the name would be `opentelemetry-collector-contrib` whereas the full name is `open-telemetry/opentelemetry-collector-contrib`.

File: receiver/gitproviderreceiver/metadata.yaml
Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml:77]: Unless I'm misunderstanding, I believe it's so you can see roughly how long your average PR takes to get merged in a historical view.

File: receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go
Comment (kentquirk) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:210]: http semconv just forced request duration to seconds from ms, so it seems like using seconds is the right answer.

Comment (4lch4) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go]: The latest commits address this and we now emit seconds instead of hours.

File: .chloggen/more-github-metrics.yaml
Comment (adrielp) [inline .chloggen/more-github-metrics.yaml:19]: We don't need to callout that it's the third pr. We can just leave the list of metrics.

File: .chloggen/more-github-metrics.yaml
Comment (adrielp) [inline .chloggen/more-github-metrics.yaml:26]: This metric has been removed per our pairing session and should be removed here.

File: receiver/gitproviderreceiver/documentation.md
Comment (adrielp) [inline receiver/gitproviderreceiver/documentation.md:107]: I think we agreed to not say `the full name` and instead just say `The name of a Git repository`.

File: receiver/gitproviderreceiver/documentation.md
Comment (adrielp) [inline receiver/gitproviderreceiver/documentation.md:93]: I think we agreed to not say `the full name` and instead just say `The name of a Git repository`.

File: receiver/gitproviderreceiver/documentation.md
Comment (adrielp) [inline receiver/gitproviderreceiver/documentation.md:78]:
```suggestion
| repository.name | The name of a Git repository | Any Str |
```

File: receiver/gitproviderreceiver/metadata.yaml
Comment (adrielp) [inline receiver/gitproviderreceiver/metadata.yaml]:
```suggestion
    description: The name of a Git repository
```

Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml]: This was just resolved, not sure how this one slipped through the cracks.

File: .chloggen/more-github-metrics.yaml
Comment (adrielp) [inline .chloggen/more-github-metrics.yaml:10]:
```suggestion
note: Add pull request metrics
```

File: .chloggen/more-github-metrics.yaml
Comment (adrielp) [inline .chloggen/more-github-metrics.yaml:19]:
```suggestion
```

File: .chloggen/more-github-metrics.yaml
Comment (adrielp) [inline .chloggen/more-github-metrics.yaml:20]:
```suggestion
```

File: .chloggen/more-github-metrics.yaml
Comment (adrielp) [inline .chloggen/more-github-metrics.yaml:26]:
```suggestion
```

File: .chloggen/more-github-metrics.yaml
Comment (astencel-sumo) [inline .chloggen/more-github-metrics.yaml:22]:
```suggestion
  - git.repository.pull_request.merged.time
```

File: .chloggen/more-github-metrics.yaml
Comment (adrielp) [inline .chloggen/more-github-metrics.yaml:23]:
```suggestion
  - git.repository.pull_request.approved.time
```

File: receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go
Comment (astencel-sumo) [inline receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:142]: What about PRs that were not merged but were instead closed without merging? Are they counted as open here?

Comment (astencel-sumo) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:142]: Oh I think I see, in `helpers.go` we're only retrieving Open and Merged PRs, so we disregard Closed PRs.

Comment (4lch4) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:142]: I see this is marked as resolved, but just in case, that's correct. We only deal with Open and Merged PRs.

**Review by adrielp — APPROVED** [review]
Approved. Just need to get the workflows to run, pull in latest from main, and get 1 more approval.

**Review by atoulme — APPROVED** [review]

**Comment by atoulme** [comment]
```
Vulnerability #1: GO-2023-2402
    Man-in-the-middle attacker can compromise integrity of secure channel in
    golang.org/x/crypto
  More info: https://pkg.go.dev/vuln/GO-2023-2402
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.16.0
    Fixed in: golang.org/x/crypto@v0.17.0
    Example traces found:
Error:       #1: libpod_client.go:22:30: podmanreceiver.init calls fmt.Errorf, which eventually calls ssh.channel.Close
Error:       #2: podman_connection.go:25:2: podmanreceiver.init calls ssh.init, which eventually calls ssh.channel.Read
Error:       #3: libpod_client.go:22:30: podmanreceiver.init calls fmt.Errorf, which eventually calls ssh.channel.Write
```
This is fixed in main and you can pick up the fix by rebasing. This shouldn't stop a merge as this is not part of the required build.

**Comment by 4lch4** [comment]
I just went ahead and ran update to merge changes from `main`.

File: receiver/gitproviderreceiver/metadata.yaml
Comment (astencel-sumo) [inline receiver/gitproviderreceiver/metadata.yaml:70]: I'm not convinced that `open` is a good namespace to create in the `git.repository.pull_request` namespace, especially if we drop `git.repository.pull_request.open.count` in favor of `git.repository.pull_request.count`.

How about `git.repository.pull_request.open_time`?

Actually, this component is in `development` status, so we don't need to sweat the naming too hard, we can always introduce breaking changes. 😉

File: receiver/gitproviderreceiver/metadata.yaml
Comment (astencel-sumo) [inline receiver/gitproviderreceiver/metadata.yaml:84]: I would get rid of the `approved` namespace. How about `git.repository.pull_request.time_to_approval`?

File: receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go
Comment (astencel-sumo) [inline receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:146]: So depending on the pull request's status (merged or not), some metrics are emitted and some are not:
- `git.repository.pull_request.merged.time` is only emitted if the pull request is merged - I guess this makes sense 😄 
- `git.repository.pull_request.open.time` is only emitted if the pull request is open - I'm not sure about this
- `git.repository.pull_request.approved.time` is only emitted if the pull request is approved and still open - I'm not sure about this either.

Anyway, this component is in `development` stability, so I guess we should just merge, evaluate how this can be used, and adjust accordingly.

Can you describe the way(s) you intend to use (or are using) these metrics? That would help a lot reasoning what makes sense.

**Review by andrzej-stencel — APPROVED** [review]
Probably worth adjusting the metric units to conform with the [spec](https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/general/metrics.md#instrument-units).

I have some comments regarding metric names and the actual usage of the metrics, but since the component is in `development` stability, I suppose the best way to iron out the details is to merge and improve incrementally.

**Comment by astencel-sumo** [comment]
Wait, I think we also need an update to the README, at least to signify what metrics are available.

File: receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go
Comment (4lch4) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:146]: > - `git.repository.pull_request.open.time` is only emitted if the pull request is open - I'm not sure about this
> - `git.repository.pull_request.approved.time` is only emitted if the pull request is approved and still open - I'm not sure about this either.

@adrielp and I talked about these when we redid the metrics last week and made this a conscious choice, but now I'm leaning more toward your take that it might be useful to emit no matter what.

File: receiver/gitproviderreceiver/metadata.yaml
Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml:56]: Oooo.. I quite like that and hadn't thought of it. I'll put this on my TODO list and get it done ASAP.

Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml:56]: @adrielp, thoughts?

File: receiver/gitproviderreceiver/metadata.yaml
Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml:70]: Yeah, I think it'd be best to change this in another PR so we don't clog this one up too much, it's already a pretty big one.

File: receiver/gitproviderreceiver/metadata.yaml
Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml:77]: I quite like this as well, mind if we do this in the next PR or should I go ahead and fix this up?

File: receiver/gitproviderreceiver/metadata.yaml
Comment (4lch4) [reply receiver/gitproviderreceiver/metadata.yaml:84]: Same as my comment above, I agree with the change but wondering if we should do this in the next PR?

**Comment by 4lch4** [comment]
@astencel-sumo, I just ran the usual "generation" commands and `go tidy` to make sure everything is up-to-date with the changes you suggested and I applied.

File: receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go
Comment (astencel-sumo) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:146]: Understanding how these metrics are/will be used would help a lot here.

Comment (adrielp) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:146]: Sorry for the delay in response to this, I've been away for the holidays and completely mentally checked out :D 


I like the comment about using state as an attribute and having these all be under the `pull_request` namespace. 

Time here really means `age`, but time is preferred name in the specification which is why we chose it.

Merged, open, and approved are all state, and thus when they change, they're essentially a different metric.

These metrics are used to help teams understand where their bottle necks are and how they could improve if they so choose.  With pull request metrics for example, a team would be able to understand what their lead time is for getting pull requests and contributions through the their code base.

If they have a lot of pull requests in the open state, and the time of which they've been open is long, then it's very likely the team is not pairing, or they have to few people available for PR reviews. 

And that might be ok for a team. But it also might be an indicator that processes should change. They're also not metrics to be used in isolation. They're meant to be used with the other metrics to form a picture.  

We have a whole doc on leading indicators and how teams could leverage these metrics that we've been working on publishing publicly. My intent (which hasn't made it in here yet) is to link & reference that guidance on these metrics within this receiver. Additionally, I'd include a link to the demo that runs and displays the visualization of these metrics.

I'd vote for merging as-is (we know these metrics work at this time in point) and make quick small changes to the names in subsequent PR's while in development.

File: receiver/gitproviderreceiver/metadata.yaml
Comment (adrielp) [reply receiver/gitproviderreceiver/metadata.yaml:56]: I like this idea. I would prefer to see it done as a change while this is in development.

File: receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go
Comment (astencel-sumo) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:146]: Awesome, I love your answer @adrielp , thank you for this.

> Sorry for the delay in response to this, I've been away for the holidays and completely mentally checked out 

That's completely OK and the right thing to do!

> We have a whole doc on leading indicators and how teams could leverage these metrics that we've been working on publishing

This sounds great, I definitely want to read it when it's public.

> I'd vote for merging as-is (we know these metrics work at this time in point) and make quick small changes

Makes sense to me, again - given that the component is in Development stability status, I suppose that's the way to go.

One last thing I'm waiting for (after the holiday season ends for everybody 🙂) is to update the README, I'm thinking about the GitHub Metrics section.﻿
image

Comment (adrielp) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:146]: Sounds good to me! Agreed on updating that readme. Hopefully that can be updated tomorrow & we can get this merged this so we can get some small pr's open to iterate! Appreciate the thorough review on this and insight!

Comment (4lch4) [reply receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:146]: There we go, I just updated the README to fix this.

File: receiver/gitproviderreceiver/README.md
Comment (adrielp) [inline receiver/gitproviderreceiver/README.md:35]:
```suggestion
- [x] Repository pull request approved time
```

File: receiver/gitproviderreceiver/README.md
Comment (adrielp) [inline receiver/gitproviderreceiver/README.md:34]:
```suggestion
- [x] Repository pull request merged time
```

**Comment by 4lch4** [comment]
Okay, this PR is finally ready to merge. My last commits reverted the earlier suggested change where units were things like `unit: "{pull_request}" as it broke the receiver. We can add this back later, but I've reverted it for now to get this PR done.

File: receiver/gitproviderreceiver/documentation.md
Comment (astencel-sumo) [inline receiver/gitproviderreceiver/documentation.md:43]: Looks like all the metric unit fixes have been lost somewhere in the rebases/merges? See [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29642#pullrequestreview-1794151974) on this PR and the changes below that comment.

```suggestion
| s | Gauge | Int |
```

**Review by andrzej-stencel — APPROVED** [review]
:rocket: :clap: :bow:

**Comment by adrielp** [comment]
Looks like the build-and-test for windows is failing because it's trying to test a podman receiver on windows. This shouldn't block the merge of this pr correct @atoulme @astencel-sumo ?

**Comment by adrielp** [comment]
Have resolved several conflicts after changes to main and changes for lifecycle tests being merged. Workflows have been re-run now and the unrelated windows workflows that were previously failing look to be fixed. We're good to go!

**Pull request merged**
